### PR TITLE
Parallelising CI based tests

### PR DESCRIPTION
**Pull request by rasswanth-s**
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: .github/workflows/pr-tests-syft.yml [hunk 1]
```diff
@@ -30,6 +30,14 @@ jobs:
 
     runs-on: ${{ matrix.os }}
     steps:
+      - name: "clean .git/config"
+        if: matrix.os == 'windows'
+        continue-on-error: true
+        shell: bash
+        run: |
+          echo "deleting ${GITHUB_WORKSPACE}/.git/config"
```
Comment (koenvanderveen) [inline .github/workflows/pr-tests-syft.yml:38]: why is this needed?

**Comment by koenvanderveen** [comment]
It would also be very helpful if we could partition the unit tests among different runners with something like this
https://stackoverflow.com/questions/56698210/limit-the-number-of-test-cases-to-be-executed-in-pytest

**Comment by koenvanderveen** [comment]
Lastly I think it would be good to slightly reduce the nr of tests we have. We are parameterizing very heavily in some test files such as action_object_test.py, resulting in over 200 tests that do a lot of similar things

File: .github/workflows/pr-tests-syft.yml (hunk 1 above)
Comment (rasswanth-s) [reply .github/workflows/pr-tests-syft.yml:38]: This a permissions on windows self hosted runners, 

Generally during multiple runs, in the docker container tests, the self hosted runners have the concept of new users for each run.

They create temporary files in the pysyft directory.

when it is used in a new run , there are permission issues while actions/checkout

https://github.com/actions/checkout/issues/861

**Comment by rasswanth-s** [comment]
> I think apart from making them parallel I believe it would also be useful to have a separate job for every folder. in #7708
> 
> I am using
> 
> ```
>         notebook-paths:
>           [
>             "tutorials/data-engineer",
>             "tutorials/data-owner",
>             "tutorials/data-scientist",
>             "tutorials/hello-syft",
>             "tutorials/pandas-cookbook",
>             "tutorials/pandas-cookbook-level-2",
>           ]
> ```

This is good idea.

I used to do this a lot while in SyMPC Repo(https://github.com/OpenMined/SyMPC)

This has a diminishing return.
Currently github allocates only a specific number of runners for each organization/repo level.

Currently,when one particular PR has a large number of runners..

This would not allow any other parallel PR's to execute the github runners.

As the other PR's have to wait until the first PR's runners have finished.

In 0.7 , we had a mix of self hosted runners (in Azure) + using github own runners

**Comment by koenvanderveen** [comment]
hmm, in that case we definitely need self hosted runners
